[GUI] Support saving image without GUI showing a window

thinking-tower opened the pull request:
Fixes #1616. cc @archibate @yuanming-hu.

# Description
As mentioned in #1616 by @ehannigan, saving the screenshot of a GUI without an actual window will throw 

`"RuntimeError: [x11.cpp:create_window@139] Taichi fails to create a window. This is probably due to the lack of an X11 GUI environment. If you are using ssh, try ssh -XY."`

It is possible to do this with `matplotlib` however it would be better to have a native solution. This PR provides the functionality to save an image without showing a GUI as well as correcting instances of `idendity`.

# Concerns
1. Since I'm putting a `if` statement around `self.core.update()`, is it a concern that the internal `GUI` class has a different state from `ti.GUI` class?

    Original version is shown here: https://github.com/taichi-dev/taichi/blob/6052a541bed7e2645110f21379a7266c5ce21624/python/taichi/misc/gui.py#L358-L363


2. I made a `test_gui.py` but there is also another `test_gui.py` in `misc/`. 
3. It could be better to recommend that if `show_GUI` is `False`, that users use `gui.core.screenshot("filename.jpg")` instead of altering the `show` function.

yuanming-hu commented on python/taichi/misc/gui.py:
```diff
@@ -34,14 +34,21 @@ class Event:
-    def __init__(self, name='Taichi', res=512, background_color=0x0):
+    def __init__(self,
+                 name='Taichi',
+                 show_GUI=True,
```
Let's use `show_gui` :-)

yuanming-hu commented on python/taichi/misc/gui.py:
```diff
@@ -34,14 +34,21 @@ class Event:
+        self.show_GUI = show_GUI
         self.core = ti_core.GUI(name, core_veci(*res))
+        if self.show_GUI:
+            self.core.initialise_window()
```
`initialise_window` -> `initialize_window`

yuanming-hu commented on taichi/gui/gui.h:
```diff
@@ -431,7 +431,7 @@ class Canvas {
   }
 
-  void set_idendity_transform_matrix() {
+  void set_identity_transform_matrix() {
```
Nice catch! Thanks!

yuanming-hu commented on tests/python/test_gui.py:
```diff
@@ -0,0 +1,31 @@
+        gui.set_image(pixels)
+        image_path = make_temp_file(suffix='.jpg')
+        gui.show(image_path)
+        image = np.array(Image.open(image_path))
```
We can use `ti.imread` and get rid of the dependency on PIL :-) 

https://taichi.readthedocs.io/en/latest/gui.html?highlight=imread#ti.imread

thinking-tower commented:
Hey @yuanming-hu, how do I fix this warning?

![Screen Shot 2020-09-02 at 3 34 55 PM](https://user-images.githubusercontent.com/37529096/91929448-80023f00-ed32-11ea-95cf-e8b525770f8f.png)

Here's the relevant snippet:
```
pixels = ti.field(dtype=ti.u8, shape=(n, n, 3))

@ti.kernel
def paint(c: dtype):
    for i, j, k in pixels:
        pixels[i, j, k] = c

for i in range(255):
    paint(i)
```

archibate commented on python/taichi/misc/gui.py:
```diff
@@ -34,14 +34,18 @@ class Event:
-    def __init__(self, name='Taichi', res=512, background_color=0x0):
+    def __init__(self,
+                 name='Taichi',
+                 show_gui=True,
+                 res=512,
```
```suggestion
                 res=512,
                 show_gui=True,
```
Some examples wil break by the argument order.

archibate commented on taichi/gui/gui.h:
```diff
@@ -805,14 +815,12 @@ class GUI : public GUIBase {
 
   void update() {
     frame_id++;
-    redraw_widgets();
```
Why remove?

thinking-tower replied:
@archibate I moved it to a separate function called `update_gui` to make it cleaner. [Here](https://github.com/taichi-dev/taichi/pull/1817/files#diff-bfe811ffeab1312fe18395d9aa2a79beR847-R853), I do

```
if (show_gui) {
    update_gui();
}
```

archibate commented on taichi/gui/gui.h:
```diff
@@ -825,13 +833,21 @@ class GUI : public GUIBase {
     auto real_fps = last_frame_interval.size() /
                     (std::accumulate(last_frame_interval.begin(),
                                      last_frame_interval.end(), 0.0_f));
+    if (show_gui) {
```
In fact, we can ignore the FPS-saving `sleep` procedure when we just want to save images.

thinking-tower replied:
@archibate actually should I also move `taichi::Time::wait_until(last_frame_time + frame_delta_limit)` to under the `if (show_gui)`?

archibate replied:
Yes, I think actually we can move all the codes to `update_gui()`? Or simply add `if (!show_gui) return;` to the first line of `update()`?

archibate commented on tests/python/test_gui.py:
```diff
@@ -0,0 +1,30 @@
+        for i, j, k in pixels:
+            pixels[i, j, k] = c
+
+    gui = ti.GUI("Test", show_gui=False, res=(n, n))
```
```suggestion
    gui = ti.GUI("Test", (n, n), show_gui=False)
```

thinking-tower replied:
I think it would be a good idea to keep the named argument for `res` in case of constructor order changing?

archibate replied:
For example, `examples/stable_fluid.py` is using `ti.GUI('Stable Fluid', (res, res))`, so it's already broken by [this change](https://github.com/taichi-dev/taichi/pull/1817#discussion_r481673588). I believe many end-users are using `res` as positional argument too.
No matter if positional argument is a good design or not, we always want to enforce backward compatibility as much as possible.

thinking-tower replied:
[Here](https://github.com/taichi-dev/taichi/pull/1817/files#diff-d10040c6d1ae795e8ecdaf6616941ab5R34-R41), I've moved `show_gui` to the last argument in the constructor now so that the original order is preserved:

```
def __init__(self,
                 name='Taichi',
                 res=512,
                 background_color=0x0,
                 show_gui=True):
```

archibate reviewed (changes requested):
Also add `if (show_gui)` guard to `cocoa.cpp`?

thinking-tower commented:
> Also add if (show_gui) guard to cocoa.cpp?

Since all the OS X window resources/functions are called in `process_event`, `set_title`, `redraw`, `redraw_widgets` (which have been wrapped in `update_gui` and I've added 
```
if (show_gui) {
    update_gui();
}
```, it seems to be okay for now.

I added `if` statements for `win32.cpp` and `x11.cpp` because they are releasing resources in the destructor unlike in `cocoa.cpp`.

yuanming-hu commented on tests/python/test_gui.py:
```diff
@@ -0,0 +1,30 @@
+        else:
+            paint(i * 1.0 / n)
+        gui.set_image(pixels)
+        image_path = make_temp_file(suffix='.jpg')
```
```suggestion
        image_path = make_temp_file(suffix='.png')
```

Let's use a lossless format here. (I know jpg images with a single color, as a special case, is probably lossless, but let's try not to confuse people with special cases :-)

yuanming-hu reviewed (approved):
Awesome! LGTM now. Just a final nit. I can merge if that suggestion is adopted and CI passes.
